### DISCUSSION TITLE
CLI Roadmap

### DISCUSSION THREAD

**Discussion by alesmit** [pequehq/framework#87] [Ideas]
The CLI should be named `peque` and allow to do operations across the Peque ecosystem such as:

- Create a new project with one command
- Keep users away from manually installing framework dependencies
- Provide a one-command API to add extensions
- Take care of boilerplate code generation

Please follow the examples below for ideas of possible implementations.

## Create a new project with one command
```bash
npx peque new project-name
```
Should:
1. Create a new `project-name` folder
2. Initialise a package.json
3. Install dependencies such as express, middlewares, etc (maybe have options like `--use-npm` or `--use-yarn` to let people go with their favourite package manager)
4. Add an example controller
5. Print instructions to start the dev environment

## Keep users away from manually installing framework dependencies
Ideally we should not let users deal with dependencies such as express, swagger, etc. but have them hidden somewhere to avoid issues with different package versions and related.
Reading this article could be a starting point: https://medium.com/trabe/cli-commands-to-wrap-third-party-tools-with-ejectable-default-config-5b79366e564c
_Note: if users would want to handle package versions, they should still be able to "eject" the configuration_

## Provide a one-command API to add extensions
Let's say someone has a project and they want to add GraphQL to it. We should provide a command like:
```bash
npx peque add graphql
```
That, if launched in the project root, should add extension packages like `@pequehq/graphql` and related dependencies in one shot. Nice-to-have, but it should also update the root module's configuration to add the newly installed module.

## Take care of boilerplate code generation
The CLI should as well provide commands to automatically create controllers, modules, and whatsoever. Some examples:
```bash
# create new controller MyController and add it to the module MyModule
npx peque generate controller MyController --target-module MyModule

# create new transformer LowerCaseTransformer
npx peque generate transformer LowerCaseTransformer

# create new module Cinema and controllers Movie, Actor, Genre tied to it
npx peque generate module Cinema --with-controllers Movie,Actor,Genre
```

**Comment by ukaoskid** [comment]
I would also prompt the user some options at project creation time, like what kind of project it is going to be set up. Some examples:

```bash
What kind of project are you setting up (multiple choice)?
- [ ] REST API
- [ ] Microservice 
```

And then show options like: 
```bash
Do you want to add any extensions (multiple choice)?
- [ ] GraphQL
- [ ] SMB Client
- [ ] Redis Client
- [ ] MQTT Client
- [ ] Apache Kafka Client
```
